tests/jax/test_distributed_fused_attn.py

```diff
@@ -310,3 +320,193 @@ def ref_func(query, kv, mask):
+                mesh_resource.cp_resource,
+                mesh_resource.tp_resource,
+                None,
+            )
+            qkv_sharding = NamedSharding(mesh, qkv_ps)
+
+            mask_ps = PartitionSpec(
+                mesh_resource.dp_resource, None, mesh_resource.cp_resource, None
+            )
+            mask_sharding = NamedSharding(mesh, mask_ps)
+
+            reorder = partial(
+                reorder_causal_load_balancing, cp_size=cp_size, tensor_format=qkv_format
+            )
+            inverse_reorder = partial(
+                inverse_reorder_causal_load_balancing, cp_size=cp_size, tensor_format=qkv_format
+            )
+
+            if load_balanced:
+                q, k, v = jax.tree.map(reorder, (q, k, v))
+
+            q_, k_, v_ = map(partial(jax.device_put, device=qkv_sharding), [q, k, v])
+            mask_ = jax.device_put(mask, device=mask_sharding)
+
+            target_func_jit = jax.jit(
+                jax.value_and_grad(target_func, argnums=[0, 1, 2]),
+                in_shardings=[qkv_sharding, qkv_sharding, qkv_sharding, mask_sharding],
+                out_shardings=(None, (qkv_sharding, qkv_sharding, qkv_sharding)),
+            )
+
+            target_fwd, target_grads = target_func_jit(q_, k_, v_, mask_)
+
+            if load_balanced:
+                target_dq, target_dk, target_dv = jax.tree.map(inverse_reorder, target_grads[0:3])
+                target_grads = (target_dq, target_dk, target_dv, *target_grads[3:])
+
+            assert_allclose(target_fwd, ref_fwd, dtype=dtype)
+
+            for i in range(len(target_grads)):
+                if ref_grads[i] is None or target_grads[i] is None:
+                    # expect both none if one is
+                    assert target_grads[i] is None and ref_grads[i] is None
+                else:
+                    try:
+                        assert_allclose(target_grads[i], ref_grads[i])
+                    except AssertionError as e:
+                        print(f"target_grads[{i}] v. ref_grads[{i}]")
+                        print("min: ", jnp.min(target_grads[i]), jnp.min(ref_grads[i]))
+                        print("max: ", jnp.max(target_grads[i]), jnp.max(ref_grads[i]))
+                        print("mean: ", jnp.mean(target_grads[i]), jnp.mean(ref_grads[i]))
+                        print("median: ", jnp.median(target_grads[i]), jnp.median(ref_grads[i]))
+                        print("std: ", jnp.std(target_grads[i]), jnp.std(ref_grads[i]))
+                        print("var: ", jnp.var(target_grads[i]), jnp.var(ref_grads[i]))
+                        print("max diff: ", jnp.max(jnp.abs(target_grads[i] - ref_grads[i])))
+                        raise e
+
+
+class TestReorderCausalLoadBalancing:
+    @pytest.mark.parametrize("cp_size", [2, 4, 8])
+    @pytest.mark.parametrize("shape", [[1, 16, 1, 1], [4, 32, 12, 32], [3, 32, 8, 64]])
+    @pytest.mark.parametrize("dtype", DTYPES)
```

Comment:
We can just testing `bfloat16`. There are many configurations for the flash attention and may cause the time out on CI. The test_fused_attn.py also only tested the bfloat16 data type.

Reply:
Done.